lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpLink.java

```diff
@@ -0,0 +1,138 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.wake.remote.transport.netty;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.handler.codec.http.*;
+import org.apache.reef.wake.remote.Encoder;
+import org.apache.reef.wake.remote.transport.Link;
+import org.apache.reef.wake.remote.transport.LinkListener;
+
+import java.net.SocketAddress;
+import java.net.URI;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Link implementation with Netty.
+ *
+ * If you set a {@code LinkListener<T>}, it keeps message until writeAndFlush operation completes
+ * and notifies whether the sent message transferred successfully through the listener.
+ */
+public final class NettyHttpLink<T> implements Link<T> {
+
+  public static final int INT_SIZE = Integer.SIZE / Byte.SIZE;
+
+  private static final Logger LOG = Logger.getLogger(NettyHttpLink.class.getName());
+
+  private final Channel channel;
+  private final Encoder<? super T> encoder;
+  private final LinkListener<? super T> listener;
+  private final URI uri;
+
+  /**
+   * Constructs a link.
+   *
+   * @param channel the channel
+   * @param encoder the encoder
+   * @param uri     the URI
+   */
+  public NettyHttpLink(final Channel channel,
+                       final Encoder<? super T> encoder,
+                       final URI uri) {
+    this(channel, encoder, null, uri);
+  }
+
+  /**
+   * Constructs a link.
+   *
+   * @param channel  the channel
+   * @param encoder  the encoder
+   * @param listener the link listener
+   * @param uri      the URI
+   */
+  public NettyHttpLink(
+          final Channel channel,
+          final Encoder<? super T> encoder,
+          final LinkListener<? super T> listener,
+          final URI uri) {
+    this.channel = channel;
+    this.encoder = encoder;
+    this.listener = listener;
+    this.uri = uri;
+  }
+  /**
+   * Writes the message to this link.
+   *
+   * @param message the message
+   */
+  @Override
+  public void write(final T message) {
+    LOG.log(Level.FINEST, "write {0} :: {1}", new Object[] {channel, message});
+    try {
+      final FullHttpRequest request =
+          new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.POST, uri.getRawPath());
+      final ByteBuf buf = Unpooled.copiedBuffer(encoder.encode(message));
+      request.headers()
+          .set(HttpHeaders.Names.HOST, uri.getHost())
+          .set(HttpHeaders.Names.CONNECTION, HttpHeaders.Values.KEEP_ALIVE)
+          .set(HttpHeaders.Names.ACCEPT_ENCODING, HttpHeaders.Values.GZIP)
+          .set(HttpHeaders.Names.CONTENT_TYPE, "application/wake-transport")
+          .set(HttpHeaders.Names.CONTENT_LENGTH, buf.readableBytes());
+      request.content().clear().writeBytes(buf);
+      final ChannelFuture future = channel.writeAndFlush(request);
+      future.sync();
```

Comment:
Why sync here?

Reply:
I wanted to throw exception when the result of future is failure by `sync()`. but i found its order is weird because listener must be added before its sync!